runtime/src/iree/hal/drivers/hip/event_semaphore.c

```diff
@@ -99,484 +132,749 @@ static void iree_hal_hip_semaphore_destroy(
+    }
+  }
 
-  // Advance the deferred work queue if possible. This also must happen
-  // outside the lock to avoid nesting.
-  iree_status_t status =
-      iree_hal_deferred_work_queue_issue(semaphore->work_queue);
+  if (iree_status_is_ok(status) && !semaphore_hit) {
+    status = iree_wait_any(wait_set, deadline_ns, NULL);
+    iree_wait_set_free(wait_set);
+    if (iree_status_is_ok(status)) {
+      // Now we have to walk all of the semaphores to propagate
+      // any errors that we find.
+      for (iree_host_size_t i = 0; i < semaphore_list.count; ++i) {
+        if (iree_hal_hip_semaphore_is_aborted(semaphore_list.semaphores[i])) {
+          status = iree_make_status(IREE_STATUS_ABORTED,
+                                    "the semaphore was aborted");
+          break;
+        }
+      }
+    }
+  }
 
+  for (iree_host_size_t i = 0; i < semaphore_list.count; ++i) {
+    iree_hal_resource_release(&cpu_events[i]->resource);
+  }
+  iree_allocator_free(host_allocator, cpu_events);
   IREE_TRACE_ZONE_END(z0);
   return status;
 }
 
-static void iree_hal_hip_semaphore_fail(iree_hal_semaphore_t* base_semaphore,
-                                        iree_status_t status) {
+static iree_status_t iree_hal_hip_event_semaphore_advance(
+    iree_hal_semaphore_t* base_semaphore) {
   iree_hal_hip_semaphore_t* semaphore =
       iree_hal_hip_semaphore_cast(base_semaphore);
   IREE_TRACE_ZONE_BEGIN(z0);
 
-  const iree_status_code_t status_code = iree_status_code(status);
-
-  iree_slim_mutex_lock(&semaphore->mutex);
+  iree_hal_hip_semaphore_work_item_t* work_item = NULL;
+  iree_hal_hip_semaphore_work_item_t* last_work_item = NULL;
 
-  // Try to set our local status - we only preserve the first failure so only
-  // do this if we are going from a valid semaphore to a failed one.
-  if (!iree_status_is_ok(semaphore->failure_status)) {
-    // Previous sta-tus was not OK; drop our new status.
-    IREE_IGNORE_ERROR(status);
+  // Take out all of the values from the queue that are less than the
+  // current visible value, and make sure we advance any work needed
+  // on them.
+  do {
+    iree_slim_mutex_lock(&semaphore->mutex);
+    iree_hal_hip_util_tree_node_t* node =
+        iree_hal_hip_util_tree_first(&semaphore->event_queue.tree);
+    if (node == NULL) {
+      iree_slim_mutex_unlock(&semaphore->mutex);
+      break;
+    }
+    if (iree_hal_hip_util_tree_node_get_key(node) >
+        semaphore->current_visible_value) {
+      iree_slim_mutex_unlock(&semaphore->mutex);
+      break;
+    }
+    iree_hal_hip_semaphore_queue_item_t copy =
+        *(iree_hal_hip_semaphore_queue_item_t*)
+            iree_hal_hip_util_tree_node_get_value(node);
+    iree_hal_hip_util_tree_erase(&semaphore->event_queue.tree, node);
     iree_slim_mutex_unlock(&semaphore->mutex);
-    IREE_TRACE_ZONE_END(z0);
-    return;
-  }
+    if (copy.event) {
+      iree_hal_hip_event_release(copy.event);
+    }
```

Comment:
iree_hal_hip_event_release handles the NULL case
```suggestion
    iree_hal_hip_event_release(copy.event);
```